### DISCUSSION TITLE
Need help understanding tasks for generic workflows

### DISCUSSION THREAD

**Discussion by Atul-Lohiya** [infiniticio/infinitic#161] [Q&A] [answered]
I went through the docs and understand that a workflow class needs to call task class/service methods from the wrokflow starting method. The thing is each task should be uniquely named with a specific purpose. For e.g., in the hello world example, the HelloWorld workflow has the greet starting method which calls the sayHello and addEnthusiasm methods "statically".
Also, these workflows and tasks must be "statically" defined in the workflowConfig.yml file.

All this is really clean for static workflows - workflows which have a defined set of tasks.

My use case is this requiring dynamic workflows: users (people who want to automate their work by connecting two or more different apps) can define workflows with mutliple tasks in them from our workflow UI. This workflow and task information is saved in our separate DB. 

From user perspective, each workflow is different. But technically they are all just one dynamic workflow - a generic workflow that defines a generic task multiple times with different parameters causing it to connect to a different app/service each time. The generic task does just one thing - call a service (technically a web API) as configured by the users.

I want to understand would the same task with different parameters be run each time it is called for next task in the workflow? Note that the "next task" in Infinitic terms is not really an Infinitic task but just some data saved in our separate DB. When the workflow starts, it needs to fetch the list of tasks from DB as was defined by the user. Then it needs to run the same generic Infinitic task with different data as fetched from DB.

This should look something like this:

`interface iGenerictaskService
{
   runTask(params);
}

class genericTask extends Task implements iGenerictaskService
{
   runTask(object params)
   {
      //based on params, call a web api, when response is received, return the data
   }
}

interface iGenericWorkflowService
{
  run();
}

class GenericWorkflow extends Workflow implements iGenericWorkflow
{
       private final iGenericWorkflowService genericWorkflowTask = newTask(iGenericWorkflowService.class);

      public  object  run()
      {
         //define params variable and get its value from our custom DB. The value is the next task to execute in the workflow.

         //run task; note there is just one generic task which connects to different services simulating running of different tasks each time
         object result = genericWorkflowTask.runTask(params);

         //The problem: when the next task needs to be executed, will the above line actually run the task or just skip it as it was run already and now has a value available
      }

}`

As the docs say, once a task has been run and completed, the workflow starts again. But the task taht was already run and whose value is now known does not run again. And that is where my problem lies. Is it possible to have a loop to iterate through the various tasks defined by user and read from DB as data and invoke the same single generic task passing the task data? Would it run the task multiple times as a new task or would it run it only once and then end the workflow?

A quick response is much appreciated as it would help me make decision on how should I proceed with my choice of workflow tool.

Thanks alot!

**Comment by geomagilles** [comment, marked as answer]
Hi, yes this will work - once executed the first task won't be executed again. Internally, a task is defined by its name but also its parameters and its "position" (1st task, 2nd task, etc.).

Actually, I worked with a marketing company that does exactly that and uses Infinitic as the workflow engine of their automation service.

The way I would do what (in pseudo code):

```
class GenericWorkflow extends Workflow implements iGenericWorkflow
{
  private genericTask = newTask(iGenericTaskService.class);
  private workflowDefinitionTask = newTask(iWorkflowDefinitionService.class);

  public  object  run(userId)
  {
     workflowDefinition = workflowDefinitionTask.get(userId)

    for (taskDefinition in workflowDefinition) {
         genericTask.run(taskDefinition.params);
    }
}
```

You can do also (more readable to me) - and it lets you use special task such as `timer`:

```
class GenericWorkflow extends Workflow implements iGenericWorkflow
{
  private emailService = newTask(iEmailService.class);
  private apiA = newTask(iApiAService.class);
  private apiB = newTask(iApiBService.class);
  // ...
  private workflowDefinitionTask = newTask(iWorkflowDefinitionService.class);

  public  object  run(userId)
  {
     workflowDefinition = workflowDefinitionTask.get(userId)

    for (taskDefinition in workflowDefinition) {
        switch(taskDefinition.type) {
            SEND_EMAIL -> emailService.send(taskDefinition.params)
            WAIT_DURATION -> timer(Duration.of(taskDefinition.duration)).await()
            API_A -> apiA.call(taskDefinition.params)
            API_B -> apiB.call(taskDefinition.params)
           // ...
    }
}
```
You are free to use whatever structure in `workflowDefinition`. 

I hope it helps.

**Reply by Atul-Lohiya** [reply]
Thanks for the reply. Can you point me to an example or documentation where it explains how to create the workflow definition? Is it created statically in code or using JSON/XML or some other form of configuration? Can this definition be created dynamically?

Also, the main branch at https://github.com/infiniticio/infinitic seems to have Kotlin code only. Can you please also help me with Java code for Infinitic?

**Reply by geomagilles** [reply]
Hi, on [infinitic repo](https://github.com/infiniticio) you'll find some examples of projects: I recommend looking at those named `infinitic-example-java-*`

**Reply by geomagilles** [reply]
the workflow definition is really up to you - if you do not see how to proceed after having looked at the examples - please let me know

**Reply by Atul-Lohiya** [reply]
I will take a look at those examples. However, the repo only seems to have kotlin code. Is there also a different repo or branch than has the Inifinitic source written in Java?

**Reply by geomagilles** [reply]
Infinitic itself is coded in Kotlin, but you can use it in Java. For examples,  all repos named `infinitic-example-java-*` are in java
